### ISSUE TITLE
dvsim: QSTA support - sim_tops needs to be moved to run_opts

### ISSUE THREAD

**Issue by svenka3** [lowRISC/opentitan#4340]
Hello,
  I now have a compile-ready, sim-starting Questa in dvsim flow :-)

I had to change the **build_opts** inside: **common_sim_cfg.hjson** 

This is so because, in Questa 2-command flow, vlog-vsim, vlog is a compiler and not an elaborator. So top level needs to be passed to vsim and not vlog. Going by dvsim's build_opts, I had to move the {sim_opts} to run_opts. Code snippet below.

Now - what do you folks suggest - add a separate build_opts_qsta - sounds like an ugly work-around to me, but to keep minimal changes to existing flow maybe. Ideally we perhaps need to add a common_build_opts and VCS/XLM/QSTA/RVRA build_opts - thinking loud, maybe there are other/better solutions.

Thanks
Srini

```json

// Default build and run opts
  build_opts: [// List multiple tops for the simulation
               // VW TBD - maybe we need build_opts_qsta "{sim_tops}",
               // Standard UVM defines
```

My changes are now pushed to a local fork, will share soon.

**Comment by svenka3** [comment]
https://github.com/svenka3/opentitan/tree/qsta-support

In case some of you want to see the local fixes to get Questa working. There are some code (SV) changes, will check-in soon. Test fails at 0-ns in SIM even after those fixes, will work on it later.

Cheers
Srini

**Comment by sriyerg** [comment]
VCS has a three step flow (compile, elaborate, run). We smush the compile and elaborate into a single build step. `{sim_tops}` tells the simulator what top level designs to elaborate after the compilation is done. So it is more of an elab time switch than a compile or run time switch. It must not be moved to `{run_opts}` in `common_sim_cfg.hjson`. Instead, the right fix is to move `{sim_tops}` to the individual `hw/dv/tools/dvsim/{tool}.hjson` at the appropriate place - `{run_opts}` for Questa and `{build_opts}` for all other tools. I can make a PR with this change.

**Comment by svenka3** [comment]
@sriyerg - yes that would be an elegant way to handle this, will fix my local branch in anticipation of your proposed PR.

Thanks

**Comment by svenka3** [comment]
Quick follow-up - Questa does not require "-top"

In common_sim_cfg.hjson I see:

```json
  // Top level simulation entities.
  sim_tops: ["-top {tb}"]
```

For Questa sake we need it wthout top. Perhaps we should add the -top prefix inside {tool}.hjson ?

Thanks
Srini

**Comment by sriyerg** [comment]
Yup - will take this into account.

**Comment by sriyerg** [comment]
Hi Srini, PTAL at #4366. For Questa, you can simply expand `sim_tops` within `run_opts` in `questa.hjson`. 
```
  run_opts: [
    ...,
    "{sim_tops}",
  ]
```

**Comment by svenka3** [comment]
Hi @sriyerg

  A follow-up: I noticed individual IP sim_cfg.hjson also has sim_tops with "-top". I am fixing them one at a time, starting with UART.

> hw/ip/uart/dv/uart_sim_cfg.hjson:

Cheers
Srini

**Comment by svenka3** [comment]
Another quick note: I see "tb" repeated leading to elab errors in Questa. I see:

> hw/ip/uart/dv/uart_sim_cfg.hjson:
```json
  // Top level testbench name (sv module).
  tb: tb
```

I believe common_sim_cfg.hjson also has "tb", I am knocking that one for now.

**Comment by sriyerg** [comment]
> Another quick note: I see "tb" repeated leading to elab errors in Questa. I see:
> 
> > hw/ip/uart/dv/uart_sim_cfg.hjson:
> 
> ```json
>   // Top level testbench name (sv module).
>   tb: tb
> ```
> 
> I believe common_sim_cfg.hjson also has "tb", I am knocking that one for now.

That just tells what the high level testbench name is. I am not clear on what's causing a problem here. `tb` is not replicated in `common_sim_cfg.hjson`.

**Comment by sriyerg** [comment]
> Hi @sriyerg
> 
> A follow-up: I noticed individual IP sim_cfg.hjson also has sim_tops with "-top". I am fixing them one at a time, starting with UART.
> 
> > hw/ip/uart/dv/uart_sim_cfg.hjson:
> 
> Cheers
> Srini

Already fixed in #4366.
